routing/router.go

```diff
@@ -1793,24 +1799,78 @@ func NewRouteRequest(source route.Vertex, target *route.Vertex,
 		Restrictions:   restrictions,
 		CustomRecords:  customRecords,
 		RouteHints:     routeHints,
+		BlindedPayment: blindedPayment,
 		FinalExpiry:    finalExpiry,
 	}
 }
 
 // validate ensures that a route request is logically sane.
 func (r *RouteRequest) validate() error {
-	if r.Target == nil {
+	var (
+		blinded    = r.BlindedPayment != nil
+		routeHints = len(r.RouteHints) != 0
+		targetSet  = r.Target != nil
+	)
+
+	// Perform some additional validation for blinded paths.
+	if blinded {
+		if targetSet {
+			return errors.New("target node can't be set " +
+				"with blinded path")
+		}
+
+		if routeHints {
+			return errors.New("route hints can't be set " +
+				"with blinded path")
+		}
+
+		if err := r.BlindedPayment.Validate(); err != nil {
+			return fmt.Errorf("invalid blinded payment: %w", err)
+		}
+
+		return nil
+	}
+
+	// If we're not sending to a blinded path, a target node is required.
+	if !targetSet {
 		return errors.New("target node required")
 	}
 
 	return nil
 }
 
+// target returns the destination node. In the case where we have a blinded
+// payment, this will be the *blinded* node ID of the final node in the blinded
+// path.
 func (r *RouteRequest) target() route.Vertex {
+	if r.BlindedPayment != nil {
+		hops := r.BlindedPayment.BlindedPath.BlindedHops
+
+		// If we're dealing with an edge-case blinded path that just
+		// has an introduction node (first hop expected to be the intro
+		// hop), then we return the unblinded introduction node as our
+		// target.
+		if len(r.BlindedPayment.BlindedPath.BlindedHops) == 1 {
+			return route.NewVertex(
+				r.BlindedPayment.BlindedPath.IntroductionPoint,
+			)
+		}
+
+		return route.NewVertex(hops[len(hops)-1].NodePub)
```

Comment:
ok so is this where we would insert & return a dummy final target for multiple blinded routes?

Reply:
Indeed! I'd think that the actually switcheroo (putting a marker pubkey as the last hop in the blinded path) would happen in `toRouteHints()` but this would be where we return that marker pubkey so that pathfinding knows where to go.